lib/routes/account.js

```diff
@@ -1026,37 +1073,88 @@ module.exports = function (
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+              /**
+               * This describes the logic behind verify_code and how it treats different requests.
+               *
+               * 1) Handling unverified email account
+               *
+               *    Check to see if `code` is the same as the account.emailCode
+               *
+               *    A) If it is, then this is a request to verify an account email. Verify email and
+               *       tokens associated with sessions.
+               *
+               *    B) Else this is a request to verify tokens, with an unverified account. `code` is a `tokenVerificationId`
+               *       Attempt to verify tokens, if `tokenVerificationId` is not found, fail request. If `tokenVerificationId`
+               *       is found, verify sessions and verify account email.
+               *
+               * 2) Handling verified email account
+               *
+               *    The request is coming from an already verified user account, it only can be a request to
+               *    verify tokens. The `code` param is a tokenVerificationId.
+               */
+              if (!account.emailVerified) {
+                if (!butil.buffersAreEqual(code, account.emailCode)) {
+                  // Attempt to verify token, could be token verification request
+                  return db.verifyTokens(code, tokenData)
+                    .then(function(){
+                      // Token verification success signifies this is a valid account, verify it
+                      log.event('tokenVerified', {email: account.email, uid: account.uid, locale: account.locale, tokenVerificationId:code})
 
-              return db.verifyEmail(account)
-                .then(mailer.sendPostVerifyEmail.bind(
-                    mailer,
-                    account.email,
-                    {
-                      acceptLanguage: request.app.acceptLanguage
-                    }
-                  )
-                )
+                      return db.verifyEmail(account)
+                        .then(function(){
+                          log.timing('account.verified', Date.now() - account.createdAt)
+                          log.event('verified', {email: account.email, uid: account.uid, locale: account.locale})
+                          log.increment('account.verified')
+
+                          // send a push notification to all devices that the account changed
+                          push.notifyUpdate(uid)
+                        })
+                    })
+                    .catch(function(err){
+                      if(err.errno === 116){
```

Comment:
It's not obvious what 116 is, who throws it or why - is there a named constant we can use here?